lms/static/scripts/frontend_apps/utils/api.ts

```diff
@@ -209,3 +212,75 @@ export function useAPIFetch<T = unknown>(
   const paramStr = recordToQueryString(params ?? {});
   return useFetch(path ? `${path}${paramStr}` : null, fetcher);
 }
+
+export type PaginatedFetchResult<T> = Omit<FetchResult<T>, 'mutate'> & {
+  /** Determines if currently loading the first page */
+  isInitiallyLoading: boolean;
+  loadNextPage: () => void;
+};
+
+/**
+ * Hook that fetches paginated data using authenticated API requests.
+ * It expects the result to include pagination info in a `pagination` prop.
+ *
+ * @param prop - Prop containing the list of results to collect.
+ * @param path - Path for API call, or null if there is nothing to fetch
+ * @param [params] - Query params for API call
+ */
+export function usePaginatedAPIFetch<
+  Prop extends string,
+  ListType extends any[],
+  FetchType extends { [P in Prop]: ListType } & { pagination: Pagination },
```

Comment:
This definition of generics is not great, because it forces you to define the three of them in order to have proper type inference when using the hook. However, this is the approach where I was able to do the lowest amount of inline type castings here.

I'll give it a thought separately, as ideally, the only generic type that should be required is `FetchType`, and based on that it should infer which are the valid props that have array type, and therefore, what's the `ListType`, which should match the type of that prop in `FetchType`.

Reply:
The current code seems like a good first step. If the types can be streamlined in future so that the caller only has to provide one generic that would be better.